File: peaky_finders/src/peaky_finders/core/preset/paths.py
"""Filesystem paths for presets, Skadi cache, and viewshed workspaces."""

from __future__ import annotations

import os
from pathlib import Path


def repo_root() -> Path:
    """Workspace root: monorepo parent when ``peaky_home/`` or ``projects/`` exists, else package root."""
    pkg_root = Path(__file__).resolve().parents[4]
    workspace = pkg_root.parent
    if (workspace / "peaky_home").is_dir() or (workspace / "projects").is_dir():
        return workspace
    return pkg_root


def peaky_home() -> Path:
    """Runtime home directory (``PEAKY_HOME``, else ``<repo>/peaky_home`` when present, else :func:`repo_root`)."""
    raw = os.environ.get("PEAKY_HOME", "").strip()
    if raw:
        return Path(raw).expanduser().resolve()
    workspace = repo_root()
    candidate = workspace / "peaky_home"
    if candidate.is_dir():
        return candidate.resolve()
    return workspace


def peaky_projects_dir() -> Path:
    """Project presets root (``PEAKY_PROJECTS`` or ``<peaky_home>/projects``)."""
    raw = os.environ.get("PEAKY_PROJECTS", "").strip()
    if raw:
        return Path(raw).expanduser().resolve()
    return peaky_home() / "projects"


def peaky_share_dir() -> Path:
    """Optional shared root (``PEAKY_SHARE``); default ``<peaky_home>/share``."""
    raw = os.environ.get("PEAKY_SHARE", "").strip()
    if raw:
        return Path(raw).expanduser().resolve()
    return peaky_home() / "share"


def resolved_skadi_mirror_dir() -> Path:
    """Global Skadi tile mirror (``SPLAT_CACHE`` env, else ``<peaky_home>/splat_cache``)."""
    raw = os.environ.get("SPLAT_CACHE", "").strip()
    if raw:
        return Path(raw).expanduser().resolve()
    return (peaky_home() / "splat_cache").resolve()


def ensure_skadi_mirror_dir() -> Path:
    """Like :func:`resolved_skadi_mirror_dir`, creating the directory when missing."""
    root = resolved_skadi_mirror_dir()
    root.mkdir(parents=True, exist_ok=True)
    return root


_PRESET_EXTENSIONS = frozenset({".yaml", ".yml"})


def require_preset_yaml_path(path: Path) -> None:
    """Raise if ``path`` is not accepted as a job preset filename (YAML only)."""
    suf = Path(path).suffix.lower()
    if suf == ".json":
        raise ValueError(
            f"Peaky preset paths must end with `.yaml` or `.yml` (not `{path.suffix}`); "
            f"legacy JSON presets are unsupported: {path}"
        )
    if suf not in _PRESET_EXTENSIONS:
        raise ValueError(f"preset path must end with `.yaml` or `.yml` (got suffix {path.suffix!r}): {path}")


def require_cwd_config_yaml(*, cwd: Path | None = None) -> Path:
    """Return ``./config.yaml`` in *cwd* (default ``Path.cwd()``) or raise ``FileNotFoundError``."""
    base = Path.cwd() if cwd is None else Path(cwd)
    path = (base / "config.yaml").resolve()
    if not path.is_file():
        raise FileNotFoundError(f"config.yaml not found in {base} — run peaky from your project directory")
    require_preset_yaml_path(path)
    return path


def resolved_preset_cache_dir(preset_path: Path) -> Path:
    """Per-preset serve runtime cache: ``<preset-dir>/.peaky/cache``."""
    p = Path(preset_path).expanduser().resolve()
    return (p.parent / ".peaky" / "cache").resolve()


def resolved_viewshed_root(preset_path: Path) -> Path:
    """Per-preset viewshed workspace root: ``<preset-dir>/.peaky/cache/viewsheds``."""
    return resolved_preset_cache_dir(preset_path) / "viewsheds"


def resolved_preset_slug(preset_path: Path) -> str:
    """Stable preset id for document titles."""
    path = Path(preset_path).expanduser().resolve()
    if path.stem == "config":
        return path.parent.name
    return path.stem


def resolve_preset_yaml_arg(
    raw: str | Path,
    *,
    cwd: Path | None = None,
) -> Path:
    """Resolve a CLI preset argument to an existing ``.yaml`` / ``.yml`` file."""
    arg = Path(raw).expanduser()
    base_cwd = Path.cwd() if cwd is None else Path(cwd)
    home = peaky_home()
    projects = peaky_projects_dir()

    candidates: list[Path] = []
    if arg.is_absolute():
        candidates.append(arg)
    else:
        candidates.append(base_cwd / arg)
        if arg.suffix.lower() not in _PRESET_EXTENSIONS:
            candidates.append(base_cwd / f"{arg}.yaml")
            candidates.append(base_cwd / f"{arg}.yml")
        if len(arg.parts) == 1:
            slug = arg.stem if arg.suffix.lower() in _PRESET_EXTENSIONS else str(arg)
            candidates.append(projects / slug / "config.yaml")
            candidates.append(projects / slug / "config.yml")
        candidates.append(home / arg)
        if arg.suffix.lower() not in _PRESET_EXTENSIONS and len(arg.parts) == 1:
            candidates.append(home / f"{arg}.yaml")
            candidates.append(home / f"{arg}.yml")

    seen: set[Path] = set()
    for candidate in candidates:
        path = candidate.expanduser().resolve()
        if path in seen:
            continue
        seen.add(path)
        if path.is_file():
            require_preset_yaml_path(path)
            return path

    return (candidates[0] if candidates else arg).expanduser().resolve()

File: crates/peaky-preset/src/paths.rs
//! Filesystem paths for presets, Skadi cache, and viewshed workspaces.

use std::env;
use std::path::{Path, PathBuf};

/// Workspace root (`peaky-finders-v5`).
pub fn workspace_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .ancestors()
        .nth(2)
        .expect("workspace root")
        .to_path_buf()
}

/// Runtime home directory (`PEAKY_HOME`, else `../../ops/peaky_home` when present, else workspace).
pub fn peaky_home() -> PathBuf {
    if let Ok(raw) = env::var("PEAKY_HOME") {
        let trimmed = raw.trim();
        if !trimmed.is_empty() {
            return expand_user(trimmed);
        }
    }

    let workspace = workspace_root();
    let ops_home = workspace.join("../../ops/peaky_home");
    if ops_home.is_dir() {
        return canonicalize_lossy(&ops_home);
    }

    let local = workspace.join("peaky_home");
    if local.is_dir() {
        return canonicalize_lossy(&local);
    }

    workspace
}

/// Project presets root (`PEAKY_PROJECTS` or `<peaky_home>/projects`).
pub fn peaky_projects_dir() -> PathBuf {
    if let Ok(raw) = env::var("PEAKY_PROJECTS") {
        let trimmed = raw.trim();
        if !trimmed.is_empty() {
            return expand_user(trimmed);
        }
    }
    peaky_home().join("projects")
}

/// Skadi HGT mirror for a project: ``<project>/.peaky/cache/skadi`` (``.hgt.gz`` tiles).
///
/// When ``SPLAT_CACHE`` is set, it overrides the project path (shared mirror or migration).
pub fn resolved_skadi_mirror_dir_for_project(project_dir: impl AsRef<Path>) -> PathBuf {
    if let Some(global) = splat_cache_from_env() {
        return global;
    }
    canonicalize_lossy(
        &project_dir
            .as_ref()
            .join(".peaky/cache/skadi"),
    )
}

/// Legacy global Skadi mirror when no project context (``SPLAT_CACHE`` or ``<peaky_home>/splat_cache``).
pub fn resolved_skadi_mirror_dir() -> PathBuf {
    if let Some(global) = splat_cache_from_env() {
        return global;
    }
    canonicalize_lossy(&peaky_home().join("splat_cache"))
}

fn splat_cache_from_env() -> Option<PathBuf> {
    env::var("SPLAT_CACHE")
        .ok()
        .map(|raw| raw.trim().to_string())
        .filter(|s| !s.is_empty())
        .map(|s| expand_user(&s))
}

/// Resolve a project argument to that project's directory.
///
/// Accepts a directory with `config.yaml`, a path to `config.yaml`, or a slug
/// under `$PEAKY_HOME/projects/<slug>/`.
pub fn resolve_project_dir(project: &str) -> PathBuf {
    let trimmed = project.trim();
    let path = expand_user(trimmed);
    if path.is_file() {
        return path
            .parent()
            .map(canonicalize_lossy)
            .unwrap_or_else(|| path.clone());
    }
    if path.join("config.yaml").is_file() {
        return canonicalize_lossy(&path);
    }
    if looks_like_filesystem_path(&path) {
        return canonicalize_lossy(&path);
    }
    canonicalize_lossy(&peaky_projects_dir().join(trimmed))
}

fn looks_like_filesystem_path(path: &Path) -> bool {
    path.is_absolute()
        || path.starts_with(".")
        || path.components().count() > 1
}

/// Resolve a project argument to `config.yaml`.
pub fn resolve_preset_path(project: &str) -> PathBuf {
    resolve_project_dir(project).join("config.yaml")
}

/// Per-preset serve runtime cache: `<preset-dir>/.peaky/cache`.
pub fn resolved_preset_cache_dir(preset_path: impl AsRef<Path>) -> PathBuf {
    let p = preset_path.as_ref();
    canonicalize_lossy(
        &p.parent()
            .unwrap_or_else(|| Path::new("."))
            .join(".peaky/cache"),
    )
}

/// Per-preset viewshed workspace root: `<preset-dir>/.peaky/cache/viewsheds`.
pub fn resolved_viewshed_root(preset_path: impl AsRef<Path>) -> PathBuf {
    resolved_preset_cache_dir(preset_path).join("viewsheds")
}

/// Stable preset id for document titles.
pub fn resolved_preset_slug(preset_path: impl AsRef<Path>) -> String {
    let path = preset_path.as_ref();
    if path.file_stem().and_then(|s| s.to_str()) == Some("config") {
        return path
            .parent()
            .and_then(|p| p.file_name())
            .and_then(|s| s.to_str())
            .unwrap_or("preset")
            .to_string();
    }
    path.file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("preset")
        .to_string()
}

fn expand_user(raw: &str) -> PathBuf {
    if raw.starts_with('~') {
        if let Ok(home) = env::var("HOME") {
            return PathBuf::from(home).join(raw.trim_start_matches("~/").trim_start_matches('~'));
        }
    }
    PathBuf::from(raw)
}

fn canonicalize_lossy(path: &Path) -> PathBuf {
    path.canonicalize().unwrap_or_else(|_| path.to_path_buf())
}
